internal/utils/nodeGroup.go

```diff
@@ -23,90 +29,144 @@ func CalculateNodeGroup(ctx context.Context, nodes v1.NodeList, config danav1alp
+		config.Status.ReservedResources = append(config.Status.ReservedResources, reservedResources)
+		return
+	}
+	config.Status.ReservedResources = append(config.Status.ReservedResources, danav1alpha1.ReservedResources{
+		NodeGroup: nodeGroupName,
+		Resources: debt,
+		Timestamp: metav1.Now(),
+	})
+	logr.Info(fmt.Sprintf("Added ReservedResources to nodeGroup %s", nodeGroupName))
+}
+
+// removeReservedfromConfig removes the reserved resources for a node group from the NodeQuotaConfig.
+// It takes the node group name and the NodeQuotaConfig to modify.
+func removeReservedfromConfig(nodeGroupName string, config *danav1alpha1.NodeQuotaConfig) {
+	index := -1
+	for i, reservedResources := range config.Status.ReservedResources {
+		if reservedResources.NodeGroup == nodeGroupName {
+			index = i
+		}
+	}
+	if index == -1 {
+		return
 	}
+	config.Status.ReservedResources = slices.Delete(config.Status.ReservedResources, index, index+1)
 }
 
-func ProcessSecondaryRoot(ctx context.Context, r client.Client, secondaryRoot danav1alpha1.NodeGroup, config *danav1alpha1.NodeQuotaConfig, rootSubnamespace string) (error, v1.ResourceList) {
+// ProcessSecondaryRoot processes a secondary root node group and updates the corresponding Subnamespace object and add reserved resources to the config if needed.
+// It takes a context, a client for making API requests, the secondary root node group, the NodeQuotaConfig,
+// the root subnamespace, and a logger for logging informational messages.
+// It returns an error (if any occurred) and the updated Subnamespace object (danav1.Subnamespace).
+func ProcessSecondaryRoot(ctx context.Context, r client.Client, secondaryRoot danav1alpha1.NodeGroup, config *danav1alpha1.NodeQuotaConfig, rootSubnamespace string, logr logr.Logger) (error, danav1.Subnamespace) {
 	sns := danav1.Subnamespace{}
 	if err := r.Get(ctx, types.NamespacedName{Namespace: rootSubnamespace, Name: secondaryRoot.Name}, &sns); err != nil {
-		return err, v1.ResourceList{}
+		logr.Error(err, fmt.Sprintf("Error getting the subnamespace %s", secondaryRoot.Name))
+		return err, sns
 	}
-	err, plus, debt := CalculateSecondaryRootPlusDebt(ctx, r, sns, secondaryRoot, config)
+	err, groupResources := CalculateSecondaryNodeGroup(ctx, r, secondaryRoot, config)
 	if err != nil {
-		return err, v1.ResourceList{}
+		return err, sns
 	}
-	AddReservedToConfig(debt, secondaryRoot.Name, config)
-	sns.Spec.ResourceQuotaSpec.Hard = FillterUncontrolledResources(plus, config.Spec.ControlledResources)
-	if err := r.Update(ctx, &sns); err != nil {
-		return err, v1.ResourceList{}
+	groupReserved := getReservedResourcesByGroup(secondaryRoot.Name, *config)
+	if isGreaterThan(sns.Spec.ResourceQuotaSpec.Hard, groupResources) {
+		// Nodes removed
+		debt := subtractTwoResourceList(sns.Spec.ResourceQuotaSpec.Hard, groupResources)
+		if groupReserved.NodeGroup == "" || hoursPassedSinceDate(groupReserved.Timestamp) < int(config.Spec.ReservedHoursTolive) {
```

Comment:
Put ` hoursPassedSinceDate(groupReserved.Timestamp) < int(config.Spec.ReservedHoursTolive)` in a different function. Also the `int` here is a redundant conversion